packages/kit/src/runtime/client/client.js

```diff
@@ -0,0 +1,1231 @@
+		prefetch_routes: async (pathnames) => {
+			const matching = pathnames
+				? routes.filter((route) => pathnames.some((pathname) => route[0].test(pathname)))
+				: routes;
+
+			const promises = matching.map((r) => Promise.all(r[1].map((load) => load())));
+
+			await Promise.all(promises);
+		},
+
+		_start_router: () => {
+			history.scrollRestoration = 'manual';
+
+			// Adopted from Nuxt.js
+			// Reset scrollRestoration to auto when leaving page, allowing page reload
+			// and back-navigation from other pages to use the browser to restore the
+			// scrolling position.
+			addEventListener('beforeunload', (e) => {
+				let should_block = false;
+
+				const navigation = {
+					from: current.url,
+					to: null,
+					cancel: () => (should_block = true)
+				};
+
+				callbacks.before_navigate.forEach((fn) => fn(navigation));
+
+				if (should_block) {
+					e.preventDefault();
+					e.returnValue = '';
+				} else {
+					history.scrollRestoration = 'auto';
+				}
+			});
+
+			addEventListener('visibilitychange', () => {
+				if (document.visibilityState === 'hidden') {
+					update_scroll_positions(current_history_index);
+
+					try {
+						sessionStorage[SCROLL_KEY] = JSON.stringify(scroll_positions);
+					} catch {
+						// do nothing
+					}
+				}
+			});
+
+			/** @param {Event} event */
+			const trigger_prefetch = (event) => {
+				const a = find_anchor(event);
+				if (a && a.href && a.hasAttribute('sveltekit:prefetch')) {
+					prefetch(get_href(a));
+				}
+			};
+
+			/** @type {NodeJS.Timeout} */
+			let mousemove_timeout;
+
+			/** @param {MouseEvent|TouchEvent} event */
+			const handle_mousemove = (event) => {
+				clearTimeout(mousemove_timeout);
+				mousemove_timeout = setTimeout(() => {
+					// event.composedPath(), which is used in find_anchor, will be empty if the event is read in a timeout
+					// add a layer of indirection to address that
+					event.target?.dispatchEvent(
+						new CustomEvent('sveltekit:trigger_prefetch', { bubbles: true })
+					);
+				}, 20);
+			};
+
+			addEventListener('touchstart', trigger_prefetch);
+			addEventListener('mousemove', handle_mousemove);
+			addEventListener('sveltekit:trigger_prefetch', trigger_prefetch);
+
+			/** @param {MouseEvent} event */
+			addEventListener('click', (event) => {
+				if (!router_enabled) return;
+
+				// Adapted from https://github.com/visionmedia/page.js
+				// MIT license https://github.com/visionmedia/page.js#license
+				if (event.button || event.which !== 1) return;
+				if (event.metaKey || event.ctrlKey || event.shiftKey || event.altKey) return;
+				if (event.defaultPrevented) return;
+
+				const a = find_anchor(event);
+				if (!a) return;
+
+				if (!a.href) return;
+
+				const is_svg_a_element = a instanceof SVGAElement;
+				const url = get_href(a);
+				const url_string = url.toString();
+				if (url_string === location.href) {
+					if (!location.hash) event.preventDefault();
+					return;
+				}
```

Comment:
I'm not sure I understand — why 1071 rather than here? I do think we could simplify it — `url.toString()` is just a roundabout way of saying `url.href`, and there's certainly no reason to create a new variable

```suggestion
				if (url.href === location.href) {
					if (!location.hash) event.preventDefault();
					return;
				}
```

Reply:
That'd be just for the sake of allowing the other checks to run. If there's a `rel=external` or `download` attribute but it ended up here, couldn't `preventDefault()` mess things up? Unless it'd have no effect—in which case, my bad!

Reply:
ah, I see, that makes sense. (and it turns out i already applied that suggestion locally and didn't think i'd committed it, so i'll resolve this)